src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+
+    Zion        = Int(pseudo["header"]["z_valence"])     # Pseudo-ion charge
+    rgrid       = pseudo["radial_grid"]                  # Radial grid
+    drgrid      = pseudo["radial_grid_derivative"]       # Integration coeff.s
+    lmax        = pseudo["header"]["l_max"]              # Maximum angular momentum channel
+    vloc        = pseudo["local_potential"] ./ 2         # Local potential (Ry -> Ha)
+    vloc_interp = linear_interpolation((rgrid, ), vloc)  # Interp.or for the local potential
+
+    # Kleinman-Bylander β projectors projs[l+1][i][ir]
+    # NB: UPFs store rβ, not just β
+    projs = map(0:lmax) do l
+        betas_l = filter(beta -> beta["angular_momentum"] == l, pseudo["beta_projectors"])
+        map(beta -> beta["radial_function"] ./ 2, betas_l)  # Ry -> Ha
+    end
+    projs_interp = map(projs) do projs_l
+        map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            linear_interpolation((rgrid[1:ir_cut], ), proj)
+        end
+    end
+
+    # β-projector coupling coefficients h[l+1][i,j] (also called e_{KB}, D_{ij})
+    h = Matrix[]
+    n = 1
+    for l = 0:lmax
+        nproj_l = length(projs[l+1])
+        Dij_l = pseudo["D_ion"][n:n+nproj_l-1, n:n+nproj_l-1] .* 2  # 1/Ry -> 1/Ha
+        push!(h, Dij_l)
+        n += nproj_l
+    end
+
+    # Pseudo-atomic wavefunctions
+    # Currently not used; can be used for initializing the starting wavefunction and as
+    # projectors for projected densities of states, projected wavefunctions, and DFT+U(+V).
+    pswfcs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["radial_function"], pswfcs_l)
+    end
+    pswfc_occs = map(0:lmax - 1) do l
+        pswfcs_l = filter(pseudo["atomic_wave_functions"]) do pswfc
+            pswfc["angular_momentum"] == l
+        end
+        map(pswfc -> pswfc["occupation"], pswfcs_l)
+    end
+
+    # Currently not used; can be used for initializing the starting guess density.
+    rhoatom = pseudo["total_charge_density"]
+
+    # Useful precomputed quantities (q-independent parts of Fourier transform integrands)
+    rvlocpzdr = (rgrid .* vloc .+ Zion) .* drgrid
+    r2projsdr = map(projs) do projs_l
+        map(projs_l) do proj
+            ir_cut = lastindex(proj)
+            rgrid[1:ir_cut] .* proj .* drgrid[1:ir_cut]
+        end
+    end
+
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h, pswfcs,
+           pswfc_occs, rhoatom, rvlocpzdr, r2projsdr; identifier)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space
+distance r via linear interpolation on the real-space mesh.
+
+Note: UPFs store `r[i] p_{il}(r[i])`, so r=0 is undefined and will error.
```

Comment:
p or beta?

Reply:
Can't it be computed? Or at least extrapolated? It'd be nice to handle that logic here rather than in the caller

Reply:
I think it could be extrapolated, but I wasn't sure how to get reasonable extrapolation working on the log mesh.